src/algebra/non_associative/basic.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2021 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import algebra.algebra.basic
+
+/-!
+# Non-associative, non-unital algebras
+
+This files defines non-associative, non-unital algebras together with their morphisms and basic
+lemmas to make them usable.
+
+## Main definitions
+
+  * `na_semiring`
+  * `na_algebra`
+  * `na_algebra_hom`
+
+## Tags
+
+non-associative, non-unital, algebra
+-/
+
+universes u₁ u₂ u₃ u₄
+
+variables (R : Type u₁) (A : Type u₂) (B : Type u₃) (C : Type u₄) [semiring R]
+
+set_option old_structure_cmd true
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring, or "non-associative ring" for
+short. -/
+class na_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
```

Comment:
We could at this point just open the floodgates and add all the missing ones:

```suggestion
class mul_one_class extends has_one, has_mul :=
(one_mul : ∀ (a : M), 1 * a = a)
(mul_one : ∀ (a : M), a * 1 = a)
class semigroup_with_zero extends semigroup A, mul_zero_class A
class nau_semiring extends add_comm_monoid A, distrib A, mul_zero_class A
class nu_semiring extends nau_semiring A, semigroup_with_zero A
class na_semiring extends nau_semiring A, mul_one_class A
```

Then the existing `algebra` definition could happily require `na_semiring` with minimal breakage. Supporting `nau_semiring` requires the new class you add in this PR.

Reply:
This may well be the right approach. For one thing, we will eventually want to talk about unital non-assoc algebras.

I'm biased to push ahead the the MVP and maybe return to this but perhaps now is the time.